### ISSUE TITLE
Always getting PortNotBindYet

### ISSUE THREAD

**Issue by sergeda** [wefoxplatform/rustainers#18]
This code always fails with PortNotBindYet(Port(21)) error.
I added sleep to check if it help - no luck.
docker ps shows that container is up and ports are open
```
    let runner = Runner::auto().unwrap();
    let container = runner.start(PureFtpD{
        ftp_loggin: ftp_login.to_string(),
        ftp_password: ftp_password.to_string(),
        main_port: ExposedPort::new(21),
        passive_port: ExposedPort::new(30000),
    }).await.unwrap();
    thread::sleep(Duration::from_secs(30));
    let main_port = container.main_port.host_port().await.unwrap();
```
What I'm doing wrong?

**Comment by IgorLaborieWefox** [comment]
Hi, can you share your `ToRunnableContainer` implementation?

**Comment by IgorLaborieWefox** [comment]
OK, I think that you forgot to declare your ports with `with_port_mappings([...])`.

This example works (on my laptop):

```rust
use std::thread;
use std::time::Duration;

use tracing::{info, Level};

use rustainers::runner::Runner;
use rustainers::{
    ExposedPort, ImageName, RunnableContainer, RunnableContainerBuilder, ToRunnableContainer,
    WaitStrategy,
};

mod common;
pub use self::common::*;

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    init_tracing(Level::INFO);

    let runner = Runner::auto()?;

    let image = PureFtpD::new("toto", "p@ssw0rd!");
    let container = runner.start(image).await.unwrap();
    // thread::sleep(Duration::from_secs(30));
    let main_port = container.main_port.host_port().await.unwrap();
    info!("Main port is {main_port}");
    Ok(())
}

#[derive(Debug)]
struct PureFtpD {
    ftp_loggin: String,
    ftp_password: String,
    main_port: ExposedPort,
    passive_port: ExposedPort,
}

impl PureFtpD {
    fn new(login: impl Into<String>, password: impl Into<String>) -> Self {
        Self {
            ftp_loggin: login.into(),
            ftp_password: password.into(),
            main_port: ExposedPort::new(21),
            passive_port: ExposedPort::new(30000),
        }
    }
}

impl ToRunnableContainer for PureFtpD {
    fn to_runnable(&self, builder: RunnableContainerBuilder) -> RunnableContainer {
        builder
            .with_image(ImageName::new("corpusops/pureftpd"))
            .with_wait_strategy(WaitStrategy::scan_port(21))
            .with_env([
                ("FTP_USER_NAME", &self.ftp_loggin),
                ("FTP_USER_PASS", &self.ftp_password),
            ])
            .with_port_mappings([self.main_port.clone(), self.passive_port.clone()])
            .build()
    }
}
```

Note that I've just released the `v0.12.0` version, you can use more wait strategy like `WaitStrategy::scan_port(21)`.
Therefore you don't need the `thread::sleep`

**Comment by sergeda** [comment]
Wow, thank you. Didn't expect response so fast . I'll try your code and new version.
